### PR TITLE
GP11 : TEE_SetOperationKey/2 panic reason

### PR DESCRIPTION
**Pull request by cedric-chaumont-st-dev**
Signed-off-by: Cedric Chaumont <cedric.chaumont@st.com>

### CODE REVIEW

File: lib/libutee/tee_api_operations.c [hunk 1]
```diff
@@ -444,18 +451,24 @@ TEE_Result TEE_SetOperationKey2(TEE_OperationHandle operation,
...
 	uint32_t key_size = 0;
 
-	if (operation == TEE_HANDLE_NULL)
-		TEE_Panic(0);
+	if (operation == TEE_HANDLE_NULL) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
 	/* Two keys not expected */
```
Comment (jbech-linaro) [inline lib/libutee/tee_api_operations.c]: I can see that this hasn't been changed by you, but is this really correct? Reading the spec it says:
```
TEE_HANDLE_FLAG_EXPECT_TWO_KEYS : Set if the algorithm expects two keys to be set, using
TEE_SetOperationKey2 . This happens only if algorithm is set to TEE_ALG_AES_XTS .
```

And in `TEE_SetOperationKey2` one can read:
```
The TEE_SetOperationKey2 function initializes an existing operation with two keys. This is used only for
the algorithm TEE_ALG_AES_XTS .
```

**Comment by jbech-linaro** [comment]
Just one comment, other than that:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by cedric-chaumont-st-dev** [comment]
Updated.

**Comment by cedric-chaumont-st-dev** [comment]
Updated.

File: lib/libutee/tee_api_operations.c [hunk 2]
```diff
@@ -464,97 +476,119 @@ TEE_Result TEE_SetOperationKey2(TEE_OperationHandle operation,
...
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
-		/* Supplied key has to meet required usage */
-		if ((key_info1.objectUsage & operation->info.
-		     requiredKeyUsage) != operation->info.requiredKeyUsage) {
-			TEE_Panic(0);
-		}
+	res = TEE_GetObjectInfo1(key1, &key_info1);
```
Comment (pascal-brand-st-dev) [inline lib/libutee/tee_api_operations.c]: Here, `key1` can be `TEE_HANDLE_NULL`. So it may be wrong I think.

Comment (cedric-chaumont-st-dev) [inline lib/libutee/tee_api_operations.c]: @pbrand: the key1/key2 check should prevent from that (panic). what do you think ?

**Comment by cedric-chaumont-st-dev** [comment]
Rebased. Updated.

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by cedric-chaumont-st-dev** [comment]
Squashed. Rebased. Tested (STM, ARM Juno).
28701 subtests of which 0 failed
687 test cases of which 0 failed
0 test case was skipped
TEE test application done!

File: lib/libutee/tee_api_operations.c (hunk 2 above)
Comment (seungeun2) [inline lib/libutee/tee_api_operations.c]: It looks like when key1 key2 are both TEE_HANDLE_NULL, it leads to panic(0) at the end of the function. But spec specifies one of panic reasons as follows.
 "key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object."
To me, it looks like the function shouldn't panic when both key1 and key2 are TEE_HANDLE_NULL.

**Comment by seungeun2** [comment]
It looks like when key1 key2 are both TEE_HANDLE_NULL, it leads to panic(0) at the end of the function. But spec specifies one of panic reasons as follows.
"key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object."
To me, it looks like the function shouldn't panic when both key1 and key2 are TEE_HANDLE_NULL.

**Comment by cedric-chaumont-st-dev** [comment]
"• key1 and/or key2 are not initialized." : both  key1 and  key2 have to be initialized (different from TEE_HANDLE_NULL) otherwise Panic.
"• key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object" : key1 and key 2 are both initialized (different from TEE_HANDLE_NULL) but both key1 or key2 have to be a valid handle (from TEE_GetObjectInfo1) otherwise Panic. I hope it helps.

**Comment by seungeun2** [comment]
As far as I understand, "key1 and key2 are not both TEE_HANDLE_NULL" (which @cedric-chaumont-st-dev  paraphrased as "key1 and/or key2 are not initialized") corresponds to the case where key1 and key2 are not both TEE_HANDLE_NULL, i.e., either 'only key1 is TEE_HANDLE_NULL' or  'only key2 is TEE_HANDLE_NULL'.  The spec doesn't say anything about the case when both key1 and key2 are TEE_HANDLE_NULL.(I'm not a native English speaker, so.. correct me if i'm wrong) 

The same thing goes to TEE_SetOperationKey function. The spec specifies one of panic reasons as "key is not TEE_HANDLE_NULL and is not a valid handle on a key object". It doesn't say that TA needs to panic when key is equal to TEE_HANDLE_NULL.  But it says "If key is set to TEE_HANDLE_NULL , then the operation key is cleared."  which is applied well in this PR.
But what's clearing the operation key for once TA panics?

**Comment by cedric-chaumont-st-dev** [comment]
@seungeun2: 
"• key1 and/or key2 are not initialized." is part of the Gp internal core API specification v1.1 . if a key is not initialized then it is equal to TEE_HANDLE_NULL. If one of the 2 keys or both are equal to TEE_HANDLE_NULL then it must Panic. In any case, if one of the 2 keys is equal TEE_HANDLE_NULL it will Panic and it ensures both of them are initialized at the same time.

"But what's clearing the operation key for once TA panics?" . Could you rephrase your question please ?

**Comment by cedric-chaumont-st-dev** [comment]
There is something confusing in the spec:
"key1 and key2 MUST both be non-NULL or both NULL" vs "key1 and/or key2 are not initialized".
I am going to clarify with GP WG.

**Comment by seungeun2** [comment]
@cedric-chaumont-st-dev : 
What I meant by "But what's clearing the operation key for once TA panics?" is that when key is TEE_HANDLE_NULL, the operation key needs to be cleared right? Buf if TA panics, there's no point in clearing the operation key , cause TA's not available to use the operation~ This is why  I asked the question "But what's clearing the operation key for once TA panics?" and I think TA shouldn't panic when key is TEE_HANDLE_NULL.
I would really appreciate if you keep me updated after you clarify with GP WG.
Thank you
